### ISSUE TITLE
Compatibility with TM 0.16

### ISSUE THREAD

**Issue by pdobacz** [KrzysiekJ/abci_server#6]
This is a bit premature, but I did some attempts to have this, and would like to share/leave a trail.

`abci_server` won't work on current `develop` of Tendermint (https://github.com/tendermint/tendermint/commit/0e68638af3fcac7af3289c3c308302b56d6277d6) because:
  - `go-wire` [changed](https://github.com/tendermint/go-wire/blob/develop/CHANGELOG.md#080-jan-28-2018)
  - protobuf definitions [changed](https://github.com/tendermint/abci/blob/develop/CHANGELOG.md#0100-tbd)

I tried some ugly experiments with fixing that, but I couldn't get anything working. I think it is anyway better to wait a bit for 0.16 release.

For now, let me share the dump of my attempts. It's very ugly, I can't neither Erlang nor Protobuf properly, and didn't delve deep into the new `go-wire` (c.f. [here](https://github.com/omisego/abci_server/blob/517c28b7a72bc18c57dfc9798d721d1b04a75004/src/abci_server.erl#L100) - have no idea, if that makes sense at all) :(

Feel free do use as starting point or discard altogether:
https://github.com/KrzysiekJ/abci_server/compare/master...omisego:temp/attempt_TM_v0.16?expand=1

**Comment by KrzysiekJ** [comment]
So the changes in go-wire affect Tendermint Socket Protocol? Is this documented somewhere?

**Comment by pdobacz** [comment]
> Is this documented somewhere?

not yet, as far as I know. If I become aware of such documentation, I'll let know here.

**Comment by srmo** [comment]
Go wire format has been updated. We ran into the same problems with jABCI.
If I'm not mistaken you are referring to the changes of varint? It no longer uses the custom lengt-of-length approach but the correct, protobuf zigzag encoding.

It has been documented after stirring some confusion in ABCI server implementations after 0.16 release 
https://github.com/tendermint/abci#socket-tsp

For your reference:
https://github.com/jTendermint/jabci/issues/37
https://github.com/tendermint/abci/issues/205
https://developers.google.com/protocol-buffers/docs/encoding?csw=1#signed-integers

Be aware of two points:
* some internal implementations like the pubKey encoding for validators (in EndBlock) still use the old wire varint encoding (length-of-length)
* go-wire moved to go-amino
